### ISSUE TITLE
Fallback-servers with DoH/DoT?

### ISSUE THREAD

**Issue by iJorgen** [AdguardTeam/dnsproxy#250]
Is it possible to setup DoH or DoT on the fallback-servers or only unencrypted queries is possible?

**Comment by ameshkov** [comment]
Hi! Any type of resolver may be a fallback

**Comment by iJorgen** [comment]
I found in the code that it is possible, but you need to specify the DoH/DoT servers by using IP (since DNS-resolution don't work) so it won't work for me since I need to specify a custom URL for DoH/DoT including my configuration-id.

Will go with unencrypted DNS for my fallback-server, since it's the last chance to get an answer when upstream-servers fails.

**Comment by ameshkov** [comment]
Could it be that you mean bootstrap servers and not fallback servers?

**Comment by iJorgen** [comment]
Aaah, you are correct. It was under "Bootstrap" I read that. :-)

`// Validate the bootstrap resolver. It must be either a plain DNS resolver.
// Or a DoT/DoH resolver with an IP address (not a hostname).`

Do fallback servers use the Bootstrap servers to resolve, if all upstream-servers are failing? I mean if I specified "tls://dns.adguard-dns.com" under fallback and all upstream-servers are down. How does it resolve "dns.adguard-dns.com" then?!

**Comment by samlux04** [comment]
> Aaah, you are correct. It was under "Bootstrap" I read that. :-)
> 
> `// Validate the bootstrap resolver. It must be either a plain DNS resolver. // Or a DoT/DoH resolver with an IP address (not a hostname).`
> 
> Do fallback servers use the Bootstrap servers to resolve, if all upstream-servers are failing? I mean if I specified "tls://dns.adguard-dns.com" under fallback and all upstream-servers are down. How does it resolve "dns.adguard-dns.com" then?!

Bootstrap used to resolve all DNS hostname.
```yaml
bootstrap:
  - "tls://1.0.0.1"
  - "https://1.1.1.1/dns-query"
  - "https://8.8.8.8/dns-query"
upstream:
  - "tls://dns.adguard.com"
  - "quic://dns.nextdns.io"
fallback:
  - "tls://doh.tiar.app"
  - "https://1.1.1.1/dns-query' # just in case when other down
```

**Comment by iJorgen** [comment]
> Bootstrap always used to resolve all DNS host in upstream and fallback. Just to get an IP for DNS to connect to. All client request will use upstream or fallback. There will be a time when ttl expire for the upstream and fallback hostname. It will use bootstrap again to get the new IP.

Thanks a lot for your explanation and samples!! Now it begins to fall in place how bootstrap, fallback and upstream works together.

**Comment by iJorgen** [comment]
I just figured out why I experienced issues using Bootstrap-servers over DoH/DoT. When I tested my config and started DNSproxy from CLI it worked every time, but not during a reboot. The reason was my specified time-servers (pool.ntp.org and time.cloudflare.com) in the router needs DNS to resolve, but to resolve over DoH/DoT the router must have the correct time. Moment 22...

Now I specified a few NTP-servers using their IP-address and it boots perfect each time. Maybe not "best practice" to use specified IP-addresses, but I can live with that to get more secure DNS...
